### PR TITLE
Fix getaddrinfo() to respect socket type hints

### LINKED ISSUE #30686: getaddrinfo() does not respect socket type
**Describe the bug**
When requesting an address for listening socket using `getaddrinfo()` with `AI_PASSIVE` in the flags field, the returned address is always using socket type `SOCK_STREAM`. Following example demonstrates this:

```.c
static const char *socktype_str(int type) {
    switch (type) {
    case SOCK_DGRAM:    return "SOCK_DGRAM";
    case SOCK_STREAM:   return "SOCK_STREAM";
    case SOCK_RAW:      return "SOCK_RAW";
    default:            return "SOCK_???";
    }
}

void test_getaddrinfo_hints(void) {
    struct addrinfo *res;
    struct addrinfo hints = {
        .ai_family = AF_INET,
        .ai_socktype = SOCK_DGRAM,	/* I'm requesting UDP address */
        .ai_flags = AI_PASSIVE,
    };

    int r = getaddrinfo(NULL, "1234", &hints, &res);
    if (r) {
        printf("Error: getaddrinfo() %d\n", r);
        return;
    }

    for (struct addrinfo *ai = res; ai; ai = ai->ai_next) {
        int s = socket(ai->ai_family, ai->ai_socktype, ai->ai_protocol);
        if (s == -1)
            continue;

        /* Verify that we got a correct socket type */
        if (ai->ai_socktype != SOCK_DGRAM) {
            printf("ERROR: I got type %s\n", socktype_str(ai->ai_socktype));
        }

        if (bind(s, ai->ai_addr, ai->ai_addrlen) == 0) {
            /* Bind succesfull
             * I should have now listening UDP socket on port 1234 */
            break;
        }
        close(s);
    }
    freeaddrinfo(res);
}
```

**To Reproduce**
1. Prepare test network for QEMU `./net-setup.sh`
2. `west build -b qemu_cortex_m3 -t run`
3. Check output.

**Expected behavior**
When AI_PASSIVE flag is specified in hints, I would expect `getaddrinfo()` to return `struct sockaddr` suitable to be used with `bind()` call, as defined by POSIX.

**Impact**
Usage of `getaddrinfo()` differs from other POSIX compliant systems.

**Logs and console output**
```
-- west build: running target run
[0/1] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: cortex-m3
*** Booting Zephyr OS build zephyr-v2.4.0-2344-g4def9a386b52  ***
ERROR: I got type SOCK_STREAM
thread: started
QEMU: Terminated
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain Zephyr SDK
 - Commit SHA or Version used: zephyr-v2.4.0-2344-g4def9a386b

**Additional context**

### PR DESCRIPTION
**Pull request by SeppoTakalo**
If getaddrinfo() was called with AI_PASSIVE flag in hints,
returned address defaulted to SOCK_STREAM and IPPROTO_TCP.
Fixed so that SOCK_DGRAM leads to correct address type
that can be fed to bind() directly.

Same hard coding was visible when numerical IPv4 address string
was converted to binary. That is also fixed to respect hints now.

Fixes #30686

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

### CODE REVIEW

**Comment by pfalcon** [comment]
Thanks for looking into this. getaddrinfo() implementation is known to be incomplete (doesn't support all flags/features). Did you consider extending tests/net/socket/getaddrinfo for this case, to make sure it won't regress again in the future?

File: subsys/net/lib/sockets/getaddrinfo.c [hunk 1]
```diff
@@ -126,8 +126,13 @@ static int getaddrinfo_null_host(int port, const struct zsock_addrinfo *hints,
 		addr->sin_family = AF_INET;
 		INIT_ADDRINFO(res, addr);
 		res->ai_family = AF_INET;
-		res->ai_socktype = SOCK_STREAM;
-		res->ai_protocol = IPPROTO_TCP;
+		if (hints->ai_socktype == SOCK_DGRAM) {
+			res->ai_socktype = SOCK_DGRAM
```
Comment (jukkar) [inline subsys/net/lib/sockets/getaddrinfo.c:130]: we are missing `;` here

**Review by jukkar — COMMENTED** [review]
Looks good. As already commented, could you add a unit test for these to `tests/net/socket/getaddrinfo`

File: subsys/net/lib/sockets/getaddrinfo.c (hunk 1 above)
Comment (SeppoTakalo) [reply subsys/net/lib/sockets/getaddrinfo.c:130]: Thanks, I'm currently fixing this, and adding tests for the whole null_host resolver.

File: tests/net/socket/getaddrinfo/src/main.c [hunk 2]
```diff
@@ -269,15 +269,38 @@ void test_getaddrinfo_num_ipv4(void)
 	struct sockaddr_in *saddr;
```
Comment (jukkar) [inline tests/net/socket/getaddrinfo/src/main.c:269]: Tests look ok, but please put them in the separate commit. So one commit for the actual fix, and the other commit for tests.

Comment (SeppoTakalo) [reply tests/net/socket/getaddrinfo/src/main.c:269]: Sure.
Separated to two commits now.

**Comment by SeppoTakalo** [comment]
The `getaddrinfo_null_host()` defaulted to IPv4.
Now that I added handling of the `hints` parameter, I was not sure which one should I default with address type `AF_UNSPEC`. I'm now returning IPv4 address, but modern Unix/POSIX machines return IPv6 because they may still `bind()` and be able to serve IPv4 clients. But how does the IP stack in Zephyr behave when listening with IPv6 any-address?

**Review by jukkar — APPROVED** [review]

**Comment by jukkar** [comment]
> But how does the IP stack in Zephyr behave when listening with IPv6 any-address?

Defaulting to IPv4 is probably ok now. I think we do not return IPv4 address when listening IPv6 one.

**Comment by pfalcon** [comment]
> but modern Unix/POSIX machines return IPv6

And that causes various trouble to the innocent users. I don't have references handy, but I read about that more than once, and faced that myself too. I actually believe the pendulum swung in the opposite direction by now, and various Linux distros let the IPv6 committee do its job (promoting IPv6) and themselves doing their job (serving, not confusing, their users, and that means IPv4-by-default).

> But how does the IP stack in Zephyr behave when listening with IPv6 any-address?

We don't have that Linux feature implemented. (And I'm talking about Linux, because that's what we use as an affinity for (many) things POSIX leaves underspecified. If you have references where POSIX prescribes a particular behavior in that regard, please share them.) So, to listen to both IPv4 and IPv6, a Zephyr application needs to have 2 listening sockets (and how various samples we have in samples/net/sockets/ do).

**Review by rlubos — APPROVED** [review]
Looks good

**Pull request merged**